v2/tools/generator/internal/codegen/pipeline/add_operator_spec.go

```diff
@@ -0,0 +1,178 @@
+		types:         make(astmodel.Types),
+	}
+}
+
+func (b *operatorSpecBuilder) newEmptyOperatorSpec() astmodel.TypeDefinition {
+	name := b.idFactory.CreateIdentifier(b.resource.Name().Name()+"OperatorSpec", astmodel.Exported)
+	operatorSpecTypeName := b.resource.Name().WithName(name)
+	operatorSpec := astmodel.NewObjectType()
+
+	operatorSpecDefinition := astmodel.MakeTypeDefinition(operatorSpecTypeName, operatorSpec)
+	description := "Details for configuring operator behavior. Fields in this struct are " +
+		"interpreted by the operator directly rather than being passed directly to Azure"
+	operatorSpecDefinition = operatorSpecDefinition.WithDescription([]string{description})
+
+	return operatorSpecDefinition
+}
+
+func (b *operatorSpecBuilder) newOperatorSpecProperty(operatorSpec astmodel.TypeDefinition) *astmodel.PropertyDefinition {
+	prop := astmodel.NewPropertyDefinition(
+		astmodel.OperatorSpecProperty,
+		b.idFactory.CreateIdentifier(astmodel.OperatorSpecProperty, astmodel.NotExported),
+		operatorSpec.Name()).MakeOptional()
+	desc := "The specification for configuring operator behavior. " +
+		"This field is interpreted by the operator and not passed directly to Azure"
+	prop = prop.WithDescription(desc)
+
+	return prop
+}
+
+func (b *operatorSpecBuilder) newSecretsProperty(secretsTypeName astmodel.TypeName) *astmodel.PropertyDefinition {
+	secretProp := astmodel.NewPropertyDefinition(
+		b.idFactory.CreatePropertyName(astmodel.OperatorSpecSecretsProperty, astmodel.Exported),
+		b.idFactory.CreateIdentifier(astmodel.OperatorSpecSecretsProperty, astmodel.NotExported),
+		secretsTypeName)
+	secretProp = secretProp.WithDescription("configures where to place Azure generated secrets.")
+	secretProp = secretProp.MakeOptional()
+
+	return secretProp
+}
+
+func (b *operatorSpecBuilder) addSecretsToOperatorSpec(
+	operatorSpecDef astmodel.TypeDefinition,
+	azureGeneratedSecrets []string) astmodel.TypeDefinition {
+
+	operatorSpec, ok := astmodel.AsObjectType(operatorSpecDef.Type())
+	if !ok {
+		panic(fmt.Sprintf("OperatorSpec %q was not an ObjectType, which is impossible", operatorSpecDef.Name()))
+	}
+
+	// Create a new "secrets" type to hold the secrets
+	resourceName := b.resource.Name()
+	secretsTypeName := resourceName.WithName(
+		b.idFactory.CreateIdentifier(
+			resourceName.Name()+"OperatorSecrets",
+			astmodel.Exported))
+	secretsType := astmodel.NewObjectType()
+
+	// Add the secrets property to the operator spec
+	secretProp := b.newSecretsProperty(secretsTypeName)
+	operatorSpec = operatorSpec.WithProperty(secretProp)
+
+	for _, secret := range azureGeneratedSecrets {
+		prop := astmodel.NewPropertyDefinition(
+			b.idFactory.CreatePropertyName(secret, astmodel.Exported),
+			b.idFactory.CreateIdentifier(secret, astmodel.NotExported),
+			astmodel.SecretDestinationType)
+		prop = prop.WithDescription("indicates where the corresponding secret should be placed. If omitted, the secret will not be retrieved from Azure.")
```

Comment:
Maybe customize the description per secret?

Reply:
Included the secret name rather than saying "the corresponding secret". We could technically allow the config to specify a description and use that instead. For now I think this (simple) level of customization is sufficient though. We can expand as the need arises.